test/net/sourceforge/kolmafia/KoLAdventureValidationTest.java

```diff
@@ -71,6 +73,213 @@ public static void afterAll() {
     Preferences.saveSettingsToFile = true;
   }
 
+  @Nested
+  class PreValidateAdventure {
+
+    public void checkDayPasses(
+        KoLAdventure adventure,
+        String place,
+        String html,
+        boolean perm,
+        boolean today,
+        String alwaysProperty,
+        String todayProperty) {
+      var builder = new FakeHttpClientBuilder();
+      var cleanups =
+          new Cleanups(
+              withHttpClientBuilder(builder),
+              withProperty(alwaysProperty, perm),
+              withProperty(todayProperty, today));
+      try (cleanups) {
+        var url = "place.php?whichplace=" + place;
+        builder.client.addResponse(200, html == null ? "" : html);
+
+        boolean success = adventure.preValidateAdventure();
+
+        var requests = builder.client.getRequests();
+        if (perm || today) {
+          // If we know that we have permanent or daily access, pre-validation
+          // returns true with no requests
+          assertTrue(success);
+          assertThat(requests, hasSize(0));
+        } else if (html == null) {
+          // If we have neither permanent or daily access and none is visible
+          // on the map, pre-validation returns false with one request
+          assertFalse(success);
+          assertThat(requests, hasSize(1));
+          assertPostRequest(requests.get(0), "/place.php", "whichplace=" + place);
+        } else {
+          // If we have neither permanent or daily access but the map shows
+          // access, pre-validation returns true with one request and sets
+          // daily access
+          assertTrue(success);
+          assertTrue(Preferences.getBoolean(todayProperty));
+          assertThat(requests, hasSize(1));
+          assertPostRequest(requests.get(0), "/place.php", "whichplace=" + place);
+        }
+      }
+    }
+
+    @ParameterizedTest
+    @CsvSource({
+      "The Neverending Party, neverendingPartyAlways, _neverendingPartyToday",
+      "The Tunnel of L.O.V.E., loveTunnelAvailable, _loveTunnelToday"
+    })
+    public void checkDayPassesInTownWrong(String adventureName, String always, String today) {
+      KoLAdventure adventure = AdventureDatabase.getAdventureByName(adventureName);
+      var html = html("request/test_visit_town_wrong.html");
+      // If we have always access, we don't have today access
+      checkDayPasses(adventure, "town_wrong", html, true, false, always, today);
+      // If we don't have always access, we might today access
+      checkDayPasses(adventure, "town_wrong", html, false, true, always, today);
+      // If we don't have always or today access, we might still have today access
+      checkDayPasses(adventure, "town_wrong", html, false, false, always, today);
+      // If we don't have always or today access, we might really not have today access
+      checkDayPasses(adventure, "town_wrong", null, false, false, always, today);
```

Comment:
This block of code is the same bar strings and could be refactored into `void checkPassCombinations(String place)`.

Reply:
Done.